### DISCUSSION TITLE
Are event/record hooks transactional?

### DISCUSSION THREAD

**Discussion by patronovski** [pocketbase/pocketbase#6100] [Q&A] [answered]
So the question is basically, are event/record hooks transactional if we chain them like in SQL triggers? For example, lets consider we have users table and user_profiles table. Whenever a user is created, we want to insert a user_profile inside the user event/record hook. If the insert fails, the user should also be rolled back.

I really like the way we can work in an event-driven way with event hooks in pocketbase where some action can trigger another action. With transactional support this can be very useful to separate concerns in application design.

**Comment by ganigeorgiev** [comment, marked as answer]
By default - No (_except when a hook is triggered by the [batch Web API](https://pocketbase.io/docs/api-records/#batch-createupdateupsertdelete-records)_), but with v0.23+ you can wrap the hook chain in a transaction on your own and reassign `e.App`. Something like:

- Go:
    ```go
    app.OnRecordCreate("users").BindFunc(func(e *core.RecordEvent) error {
        originalApp := e.App
        txErr := e.App.RunInTransaction(func(txApp core.App) error {
            e.App = txApp

            // do something before the create...

            // execute the create
            err := e.Next()
            if err != nil {
                return err
            }

            // do something after the create...

            return nil
        })
        e.App = originalApp

        return txErr
    })
    ```

- JS:
    ```js
    onRecordCreate((e) => {
        let originalApp = e.app
        try {
            e.app.runInTransaction((txApp) => {
                e.app = txApp

                // do something before the create...

                // execute the create
                e.next()

                // do something after the create...
            })
        } finally {
            e.app = originalApp
        }
    }, "users")
    ```

**Reply by patronovski** [reply]
Thanks, i am happy that there is a workaround.

**Reply by muety** [reply]
Very helpful, thanks! Would be cool if there was a more convenient way to accomplish this in future releases. The most painful part here is that I'd have to pass down `txApp` as a parameter to every "service" method that performs write operations to the database. Have a look at this example:

```javascript
// utils.js
function wrapTransactional(e, fn) {
    let app = e.app
    try {
        e.app.runInTransaction((txApp) => {
            e.app = txApp
            return fn(e)
        })
    } finally {
        e.app = app
    }
}

// item.pb.js
onRecordCreateExecute((e) => {
  const { wrapTransactional } = require(`${__hooks}/utils.js`)
  const { doSomething } = require(`${__hooks}/itemOps.js`)

  wrapTransactional(e, (e) => {
    e.next()

    // without transaction:  
    doSomething(e.record)

    // with transaction
    doSomething(e.record, e.app)
  })
})

// itemOps.js
// without transaction:
function doSomething(r) {
  // ...
  $app.save(r)
}

// with transaction:
function doSomething(r, app=$app) {
  // ...
  app.save(r)
}
```

If I forget to always pass `e.app` as the second argument and instead use the global `$app`, I'd end up with a deadlock.

It would already be super helpful if there was a "global" way to retrieve the currently open transaction, if there is one. Something like:

```javascript
const app = $app.getCurrentTx() || $app
```

Do you think that would be possible?

**Reply by patronovski** [reply]
Yes, i agree. Or inject it directly in the record hooks something like:

```
onRecordCreate((e, ctx) {
   ctx.currentTransaction // retrieve current transaction from the context, if no transaction is available, return null
}
```

Or

```
onRecordCreate((e) {
   e.currentTransaction // retrieve current transaction from the event itself, if no transaction is available, return null
}
```

**Reply by ganigeorgiev** [reply]
@muety I'm open for suggestions but I don't see how (_feel free to open a new discussion_).

@patronovski I don't see the point of doing this. `e.app` already contains the transaction context.

**Reply by ganigeorgiev** [reply]
@muety One thing I forgot to mention is that specific for the JS pb_hooks we overwrite the `$app` variable inside each hook handler to `e.app`, aka. in other words if the entire problem in your case comes from the fact that you don't want to pass as argument the app instance in your utils functions, then you can move your calls in another `onRecordCreateExecute` after starting a transaction in the previous one.

**Reply by muety** [reply]
> [...]  if the entire problem in your case comes from the fact that you don't want to pass as argument the app instance

The main problem comes from the lack of a straightforward way to make hooks execute in a transaction as a whole.

> [...] you can move your calls in another onRecordCreateExecute after starting a transaction in the previous one.

Can you give an example?

**Reply by ganigeorgiev** [reply]
> The main problem comes from the lack of a straightforward way to make hooks execute in a transaction as a whole.

I'm starting to regret adding support for this as it seems that users are abusing it. Wrapping the entire hook chain in a transaction SHOULD NOT be the default behavior as it could cause performance bottlenecks and should be used only when really necessary.

> Can you give an example?

```js
onRecordCreateExecute((e) => {
    let originalApp = e.app
    try {
        e.app.runInTransaction((txApp) => {
            e.app = txApp
            e.next()
        })
    } finally {
        e.app = originalApp
    }
}, "users")

onRecordCreateExecute((e) => {
   // here e.app and $app are the same and both points to the txApp started above...
   
   ...
}, "users")
```

In any case, please open a new discussion as hijacking older ones tends to become messy and makes it difficult to understand/follow for other users that may stumble here via search.

**Reply by muety** [reply]
Thanks for the example.

Last comment on this: I think allowing hooks to run transactionally is somewhat crucial, because hooks are the only way to associate a cascade of additional, subsequent actions to the creation / update / deletion of a record.

Say you're building a "click & collect" feature on top of a product catalogue, where users can reserve products to be picked up later. It involves a `reservation` and a `product` entity. Each time a new reservation is created or deleted, the according product's _instock_ status shall be updated as well. If persisting the reservation fails, product must not be updated and, vice versa, if updating the product status fails, there must not be a dangling reservation - perfect use case for a hook running in a transaction. (Technically you could argue the status shouldn't be an explicit field in the first place, but rather be inferred "dynamically" (in the sense of database normalization), but just to give an example...).
